---
title: "DESeq2"
output: html_document
date: "2024-04-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


###Introduction: This section introduces the purpose of the document, which is to provide a guide for RNA-seq data analysis using DESeq2.

###Libraries Setup: It loads necessary libraries for the analysis, such as DESeq2 for differential expression analysis, ggplot2 for data visualization, and others.

###DESeqDataSet Construction: This part constructs a DESeqDataSet object, which is a container for the count data and metadata required for DESeq2 analysis.


```{r}
# Load required libraries
suppressPackageStartupMessages({
library(DESeq2)
library(ggplot2)
library(readxl)
library(pheatmap)
library(cowplot)
library(ggrepel)  
})

#library(vsn)

# Read count data without header
count_data <- read.csv("counts_read.txt", header=FALSE, sep = "\t", row.names = 1)
count_data <- as.matrix(count_data)
dim(count_data) # check the number of rows and column


# removal of non gene rows

sum(!grepl("_g", rownames(count_data))) # get the dim of the  non gene values (row names)
print(rownames(count_data)[!grepl("_g", rownames(count_data))])

#Show in New Window
#[1] "__no_feature"           "__ambiguous"            "__too_low_aQual"        "__not_aligned"         
#[5] "__alignment_not_unique"
#filter only the name that has "_g" in the name of gen

count_data <- count_data[grep("_g", rownames(count_data)), ]

dim(count_data)

# Create header
header_names <- c("MSK_LEAF_92", "MSK_ROOT_93", "MSK_ARIL_94", "MSK_STEM_96", "MSK_ARIL_97", "MT_ARIL_66", "MT_ARIL_67", "MT_ARIL_69")
colnames(count_data) <- header_names



# Read metadata
metadata <- read_excel("RNAseq_metadata.xlsx", sheet = 1)
metadata

rownames(metadata) <- metadata$Sample_name

#to check if the rownames of metadata is equal to colnames of count data or not.
all.equal(rownames(metadata), colnames(count_data))  
```

###DESeqDataSet Construction: This part constructs a DESeqDataSet object, which is a container for the count data and metadata required for DESeq2 analysis.

```{r}
dds <- DESeqDataSetFromMatrix(countData = count_data,
                               colData = metadata,
                               design = ~  Organ + Cultivar )
dds       

names(colData(dds))
```



###Pre-processing: It includes pre-processing steps such as pre-filtering to remove genes with zero counts to improve the accuracy of the analysis.


```{r}
keep <- rowSums(counts(dds)) >= 1
dds  <- dds[keep,]
dim(dds)

```

# save the deseq2 object 

```{r}
saveRDS(dds, "/Users/razi.m/MS_Bioinformatics/Year1/Spring_Semester_2024/Genome_analysis/genome_project_1/output/read_count/dds.rds")
```

# load the object:

```{r}
# load the object:
dds <- readRDS("/Users/razi.m/MS_Bioinformatics/Year1/Spring_Semester_2024/Genome_analysis/genome_project_1/output/read_count/dds.rds")
```


#fit a negative binomial generalized linear model to the count data stored in the DESeqDataSet object dds and check the  results of the differential expression analysis from the DESeqDataSet object 

```{r}
design(dds)
dds$Cultivar <- factor(dds$Cultivar, levels = c("Musang_King", "Monthong"))
dds$Organ <- factor(dds$Organ, levels = c("non_fruit", "fruit"))
dds <- DESeq(dds)
resultsNames(dds)

```



#### QC and data exploration 

## performs a variance-stabilizing transformation on the count data stored in a DESeqDataSet (unbiased comparisons of samples)

```{r}

normCount <- rlog(dds)
head(assay(normCount), 5)

#to Inspect the norm count distribution
hist(assay(normCount))

```

### QC and Data Exploration: It includes quality control and data exploration steps such as variance-stabilizing transformation, heatmap, correlation plot, and principal component analysis (PCA) to understand the structure of the data.
```{r}

select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("Cultivar","Organ", "Tissue")])

#heatmap
pheatmap(assay(normCount)[select,], cluster_rows=FALSE, show_rownames=T,
         cluster_cols=FALSE, annotation_col=df)

```

###correlation plot

```{r}
#correlation between the samples
cor1 <- cor(assay(normCount), method = "spearman") 

#heatmap , spearman correlation is used to calculate the similarity between the sample and the spearman Rho is a measure of correlation which is a fraction of one, so to get the distance from the regression line we subtract Rho from one as a measure of distance.

pheatmap(cor1,clustering_distance_rows = as.dist(1- cor1),
          clustering_distance_cols = as.dist(1 - cor1),
          fontsize = 7, annotation_col=df)

```


### principal component analysis (PCA) to understand the structure of the data.


```{r}
 #PCA plot for sampling_site
pca_cultivar <-plotPCA(normCount, intgroup= c("Cultivar"))   
pca_cultivar
```


```{r}
 #PCA plot for sampling_site
pca_cultivar <-plotPCA(normCount, intgroup= c("Organ"))   
pca_cultivar
```




```{r}

# pca_dats , use for ggplot to have  "Organ" site and "Cultivar" in the same plot
pca_data1 <- plotPCA(normCount, intgroup= c("Organ" , "Cultivar"),returnData=T) 

percent_var <- round(100*attr(pca_data1 ,"percentVar")) # to calculate percentage value and round it which used in ggplot as input

#ggplot for "pca_data",
plot1 <- ggplot2::ggplot(pca_data1 , aes(x = PC1, y = PC2, color = Organ, shape = Cultivar)) +  coord_fixed()+ 
  geom_point(size = 2) + theme(legend.text=element_text(size=6) ,legend.key.size = unit(4, 'mm') , legend.title = element_text(size=8))+
  xlab(paste0("PC1: ",percent_var[1],"% variance"))+  # to get the variance percentage on the plot
  ylab(paste0("PC2: ",percent_var[2],"% variance"))  
 

plot1

```
###Differential Expression Results:
Here, the results of the differential expression analysis are computed for two comparisons: "Organ_fruit_vs_non_fruit" and "Cultivar_Monthong_vs_Musang_King".

###The analysis compares gene expression levels between different conditions or groups (e.g., fruit vs. non-fruit, Monthong cultivar vs. Musang King cultivar) to identify genes that are differentially expressed.
These results typically include statistical measures such as log2 fold changes, adjusted p-values, and other metrics that indicate the significance and direction of gene expression changes.

###Saving Results to CSV Files:
Once the results are computed, they are saved as CSV (comma-separated values) files.
Saving the results allows for easy access and sharing with collaborators or for further analysis using other tools or software.
Each CSV file contains a tabular representation of the differential expression results, with rows representing individual genes and columns representing various statistical measures and annotations.

###Counting Significantly Differentially Expressed Genes:
After obtaining the results, it's common to assess the number of genes that are significantly differentially expressed.
This assessment is typically based on a predefined threshold for adjusted p-values (e.g., p < 0.05), which helps determine the statistical significance of gene expression changes.
Counting the significantly differentially expressed genes provides insight into the impact of the experimental conditions or factors being compared and helps prioritize genes for further investigation or validation.

```{r}
res_org <- results(dds, name = "Organ_fruit_vs_non_fruit")
res_cult <- results(dds, name = "Cultivar_Monthong_vs_Musang_King")


write.csv(as.data.frame(res_org),
          file="/Users/razi.m/MS_Bioinformatics/Year1/Spring_Semester_2024/Genome_analysis/genome_project_1/output/read_count/res_org.csv", row.names = T)

write.csv(as.data.frame(res_cult),
          file="/Users/razi.m/MS_Bioinformatics/Year1/Spring_Semester_2024/Genome_analysis/genome_project_1/output/read_count/res_cult.csv", row.names = T)

# how many genes are significantly DE 
length(which(res_org$padj < 0.05 ))
length(which(res_cult$padj < 0.05 ))

```


###MA Plots: MA plots are created to visualize the log fold changes versus the mean of normalized counts for each gene, which helps to identify differentially expressed genes.

```{r}
par(mfrow=c(1, 2))
plotMA(res_org, ylim=c(-12,12), 
       main="Organ_fruit_vs_non_fruit", 
       cex.main=0.9, cex.lab=0.8)
plotMA(res_cult, ylim=c(-12,12), 
       main="Cultivar_Monthong_vs_Musang_King", 
       cex.main=0.9, cex.lab=0.8)
```

###Gene Expression Visualization: Finally, gene expression is visualized using ggplot2 by plotting the log fold changes versus the -log10 of the adjusted p-values. Differentially expressed genes are highlighted in different colors based on their expression levels.


```{r}

sum(res_cult$padj<0.05, na.rm = T)
sum(is.na(res_cult$padj))
dim(res_cult) 
res_cult <- na.omit(res_cult)

Expression <- ifelse(
  res_cult$log2FoldChange < -1 & res_cult$padj < 0.05, 'blue',
  ifelse(res_cult$log2FoldChange > 1 & res_cult$padj < 0.05, 'red', 'grey'))

Expression[is.na(Expression)] <- 'grey'
names(Expression)[Expression == 'red'] <- 'Up-regulated'
names(Expression)[Expression == 'grey'] <- 'Not-significant'
names(Expression)[Expression == 'blue'] <- 'Down-regulated'

# Plot using ggplot2
ggplot(res_cult, aes(x = log2FoldChange, y = -log10(padj), color = Expression, label = rownames(res_cult))) +
  geom_point(size = 2) +
  scale_color_manual(values = c('blue', 'grey', 'red'),  # Adjust colors as needed
                     labels = c('Down-regulated', 'Not-significant', 'Up-regulated')) +  # Adjust legend labels
  geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 3 & padj < 1e-3, rownames(res_cult), "")), 
                  size = 2.5, hjust = 0, vjust = 0, fontface = "bold") +
  xlim(c(floor(min(res_cult$log2FoldChange)), ceiling(max(res_cult$log2FoldChange)))) +
  ylim(c(0, max(-log10(res_cult$padj)))) +
  labs(title = 'Cultivar_Monthong_vs_Musang_King') +
  theme_minimal() +
  theme(legend.position = 'right',
        plot.title = element_text(face = "bold")) +  # Set plot title to bold
  xlab('log2FoldChange') +
  ylab('-log10(padj)') +
  guides(color = guide_legend(title = 'Expression')) + # Adjust legend title
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "grey") +  # Add vertical cutoff lines for log2FoldChange
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "grey") +  # Add horizontal cutoff line for padj
  theme(plot.title = element_text(hjust = 0.5)) 
```
###Gene Expression Visualization: Finally, gene expression is visualized using ggplot2 by plotting the log fold changes versus the -log10 of the adjusted p-values. Differentially expressed genes are highlighted in different colors based on their expression levels.

```{r}

sum(res_org$padj<0.05, na.rm = T)
sum(is.na(res_org$padj))
dim(res_org) 
res_org <- na.omit(res_org)

Expression <- ifelse(
  res_org$log2FoldChange < -1 & res_org$padj < 0.05, 'blue',
  ifelse(res_org$log2FoldChange > 1 & res_org$padj < 0.05, 'red', 'grey'))

Expression[is.na(Expression)] <- 'grey'
names(Expression)[Expression == 'red'] <- 'Up-regulated'
names(Expression)[Expression == 'grey'] <- 'Not-significant'
names(Expression)[Expression == 'blue'] <- 'Down-regulated'

# Plot using ggplot2
ggplot(res_org, aes(x = log2FoldChange, y = -log10(padj), color = Expression, label = rownames(res_org))) +
  geom_point(size = 2) +
  scale_color_manual(values = c('blue', 'grey', 'red'),  # Adjust colors as needed
                     labels = c('Down-regulated', 'Not-significant', 'Up-regulated')) +  # Adjust legend labels
  geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 5 & padj < 1e-10, rownames(res_org), "")), 
                  size = 2.5, hjust = 0, vjust = 0, fontface = "bold") +
  xlim(c(floor(min(res_org$log2FoldChange)), ceiling(max(res_org$log2FoldChange)))) +
  ylim(c(0, max(-log10(res_org$padj)))) +
  labs(title = 'Organ_fruit_vs_non_fruit') +
  theme_minimal() +
  theme(legend.position = 'right',
        plot.title = element_text(face = "bold")) +  # Set plot title to bold
  xlab('log2FoldChange') +
  ylab('-log10(padj)') +
  guides(color = guide_legend(title = 'Expression')) + # Adjust legend title
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "grey") +  # Add vertical cutoff lines for log2FoldChange
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "grey") +  # Add horizontal cutoff line for padj
  theme(plot.title = element_text(hjust = 0.5)) 
```


# Since eggnog did not provide any gene names, the “functional annotation eggNOG.mapper.annotations” file was used to get information on high or low-expression genes

```{r}


annotations <- read.table("functional_annotation_eggNOG.emapper.annotations", comment='#', sep='\t', header=F, fill=TRUE)


```

